### ISSUE TITLE
Tesseract 4 and 5 is about  200 times slower than 3 on my Linux system.

### ISSUE THREAD

**Issue by ripefig** [tesseract-ocr/tesseract#2611]
### Environment

* **Tesseract Version**: 

`> tesseract -v`

```
tesseract 4.0.0
 leptonica-1.76.0
  libgif 5.1.4 : libjpeg 8d (libjpeg-turbo 1.5.2) : libpng 1.6.36 : libtiff 4.0.10 : zlib 1.2.11 : libwebp 0.6.1 : libopenjp2 2.3.0
 Found AVX2
 Found AVX
 Found SSE
```
`tesseract-snap -v`
```
tesseract 5.0.0-alpha-335-gae02
 leptonica-1.74.2
  libjpeg 8d (libjpeg-turbo 1.4.2) : libpng 1.2.54 : libtiff 4.0.6 : zlib 1.2.8
 Found AVX2
 Found AVX
 Found FMA
 Found SSE
```

`tesseract-ocr-eng` : `1:4.00~git30-7274cfa-1`

I used the training data from the ubuntu repos for both `tesseract` and `tesseract-snap` , since no data is provided with the snap.

* **Platform**: 
Operating System: Kubuntu 19.04
KDE Plasma Version: 5.15.4
KDE Frameworks Version: 5.56.0
Qt Version: 5.12.2
Kernel Version: 5.0.0-21-generic
OS Type: 64-bit
Processors: 4 × Intel® Core™ i7-4600U CPU @ 2.10GHz
Memory: 11.6 GiB of RAM

### Current Behavior:
It takes over a minute of 100% CPU load to scan this sentence :

![62771160-3bf82880-baa5-11e9-8d39-4d9c4381a093](https://user-images.githubusercontent.com/45201036/62841051-2b65cf00-bcac-11e9-97df-bf85ff0b09bf.png)

results for tesseract 4:
`> time tesseract -l eng  62771160-3bf82880-baa5-11e9-8d39-4d9c4381a093.png 1`

```
Tesseract Open Source OCR Engine v4.0.0 with Leptonica
Warning: Invalid resolution 0 dpi. Using 70 instead.
Estimating resolution as 143

real    1m9.096s
user    3m7.484s
sys     0m0.335s

```
Tesseract 5:

`> time tesseract-snap -l eng  62771160-3bf82880-baa5-11e9-8d39-4d9c4381a093.png 1`

```
Tesseract Open Source OCR Engine v5.0.0-alpha-335-gae02 with Leptonica
Warning: Invalid resolution 0 dpi. Using 70 instead.
Estimating resolution as 143

real    1m13.585s
user    3m16.104s
```

I tried to OCR a one page doc, but I had to exit the process. It would probably take one hour of full CPU load.

### Expected Behavior:
It shouldn't take this long to scan one sentence.

**Comment by ripefig** [comment]
The solution is to set `OMP_THREAD_LIMIT=1`
Shouldn't multithreading be disabled by default until it's fixed?

https://github.com/tesseract-ocr/tesseract/issues/898

**Comment by stweil** [comment]
I cannot reproduce your timing results on a recent Debian system:

    $ tesseract --version
    tesseract 4.0.0
     leptonica-1.76.0
      libgif 5.1.4 : libjpeg 6b (libjpeg-turbo 1.5.2) : libpng 1.6.36 : libtiff 4.0.10 : zlib 1.2.11 : libwebp 0.6.1 : libopenjp2 2.3.0
     Found AVX2
     Found AVX
     Found SSE
    $ time tesseract -l eng 62841051-2b65cf00-bcac-11e9-97df-bf85ff0b09bf.png 1
    Tesseract Open Source OCR Engine v4.0.0 with Leptonica
    Warning: Invalid resolution 0 dpi. Using 70 instead.
    Estimating resolution as 143

    real	0m0,209s
    user	0m0,497s
    sys	0m0,024s

**Comment by Shreeshrii** [comment]
Is the test image available somewhere? I would like to try it on a non-AVX system.

**Comment by Shreeshrii** [comment]
Using phototest.tif, I get better timing with `OMP_THREAD_LIMIT=1`

### tessdata

```
ubuntu@tesseract-ocr:~/TEST$ OMP_THREAD_LIMIT=1 time tesseract -l eng  phototest.tif phototest --tessdata-dir ~/tessdata
Tesseract Open Source OCR Engine v5.0.0-alpha-332-gb839 with Leptonica
Page 1
5.09user 0.02system 0:05.17elapsed 98%CPU (0avgtext+0avgdata 88064maxresident)k
0inputs+128outputs (0major+2178minor)pagefaults 0swaps

ubuntu@tesseract-ocr:~/TEST$ time tesseract -l eng  phototest.tif phototest --tessdata-dir ~/tessdata
Tesseract Open Source OCR Engine v5.0.0-alpha-332-gb839 with Leptonica
Page 1

real    0m16.518s
user    0m12.848s
sys     0m0.208s
```
### tessdata_best

```
ubuntu@tesseract-ocr:~/TEST$ OMP_THREAD_LIMIT=1 time tesseract -l eng  phototest.tif phototest --tessdata-dir ~/tessdata_best
Tesseract Open Source OCR Engine v5.0.0-alpha-332-gb839 with Leptonica
Page 1
4.43user 0.03system 0:04.48elapsed 99%CPU (0avgtext+0avgdata 56704maxresident)k
0inputs+128outputs (0major+1518minor)pagefaults 0swaps

ubuntu@tesseract-ocr:~/TEST$ time tesseract -l eng  phototest.tif phototest --tessdata-dir ~/tessdata_best
Tesseract Open Source OCR Engine v5.0.0-alpha-332-gb839 with Leptonica
Page 1

real    0m8.467s
user    0m7.992s
sys     0m0.120s
```

### tessdata_fast

```
ubuntu@tesseract-ocr:~/TEST$ OMP_THREAD_LIMIT=1 time tesseract -l eng  phototest.tif phototest --tessdata-dir ~/tessdata_fast
Tesseract Open Source OCR Engine v5.0.0-alpha-332-gb839 with Leptonica
Page 1
2.20user 0.00system 0:02.25elapsed 97%CPU (0avgtext+0avgdata 35072maxresident)k
0inputs+128outputs (0major+780minor)pagefaults 0swaps


ubuntu@tesseract-ocr:~/TEST$ time tesseract -l eng  phototest.tif phototest --tessdata-dir ~/tessdata_fast
Tesseract Open Source OCR Engine v5.0.0-alpha-332-gb839 with Leptonica
Page 1

real    0m13.106s
user    0m9.045s
sys     0m0.172s
```

**Comment by stweil** [comment]
> Is the test image available somewhere? I would like to try it on a non-AVX system.

It's given in the initial report: https://user-images.githubusercontent.com/45201036/62841051-2b65cf00-bcac-11e9-97df-bf85ff0b09bf.png.

Even without AVX it should not take more than a second.

**Comment by Shreeshrii** [comment]
Thanks @stweil .

Here are the results on my system - Linux tesseract-ocr 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:54:50 UTC 2019 ppc64le ppc64le ppc64le GNU/Linux

```
ubuntu@tesseract-ocr:~/TEST$  time tesseract -l eng 62841051-2b65cf00-bcac-11e9-97df-bf85ff0b09bf.png 1
Tesseract Open Source OCR Engine v5.0.0-alpha-332-gb839 with Leptonica
Warning: Invalid resolution 0 dpi. Using 70 instead.
Estimating resolution as 143

real    0m15.212s
user    0m9.774s
sys     0m0.186s

ubuntu@tesseract-ocr:~/TEST$ OMP_THREAD_LIMIT=1 time tesseract -l eng 62841051-2b65cf00-bcac-11e9-97df-bf85ff0b09bf.png 1
Tesseract Open Source OCR Engine v5.0.0-alpha-332-gb839 with Leptonica
Warning: Invalid resolution 0 dpi. Using 70 instead.
Estimating resolution as 143
3.09user 0.02system 0:03.11elapsed 99%CPU (0avgtext+0avgdata 88064maxresident)k
0inputs+128outputs (0major+2175minor)pagefaults 0swaps
```

```
ubuntu@tesseract-ocr:~/TEST$ tesseract -v
tesseract 5.0.0-alpha-332-gb839
 leptonica-1.78.0
  libgif 5.1.4 : libjpeg 8d (libjpeg-turbo 1.4.2) : libpng 1.2.54 : libtiff 4.0.9 : zlib 1.2.11 : libwebp 0.4.4 : libopenjp2 2.3.0

IBM POWER8 - 8 CPU, 24 GB RAM
```

**Comment by ripefig** [comment]
@stweil seems like you're have the same problem, given that your system is much more powerful.

**Comment by stweil** [comment]
Test result on an ARM system:

    real	0m1.864s
    user	0m4.778s
    sys	0m0.147s

With `export OMP_THREAD_LIMIT=1`:

    real	0m2.078s
    user	0m1.950s
    sys	0m0.099s

The results for 4.0.0 and latest Git master are similar.

**Comment by stweil** [comment]
@ripefig, your results could be explained if Tesseract cannot get 4 CPU cores. On my ARM system which has 4 cores I get a faster result with `export OMP_THREAD_LIMIT=2`:

    real	0m1.426s
    user	0m1.969s
    sys	0m0.159s

**Comment by ripefig** [comment]
```
$ time OMP_THREAD_LIMIT=1  tesseract -l eng  62771160-3bf82880-baa5-11e9-8d39-4d9c4381a093.png 1
Tesseract Open Source OCR Engine v4.0.0 with Leptonica
Warning: Invalid resolution 0 dpi. Using 70 instead.
Estimating resolution as 143

real    0m0.366s
user    0m0.346s
sys     0m0.012s

$ time OMP_THREAD_LIMIT=3  tesseract -l eng  62771160-3bf82880-baa5-11e9-8d39-4d9c4381a093.png 1
Tesseract Open Source OCR Engine v4.0.0 with Leptonica
Warning: Invalid resolution 0 dpi. Using 70 instead.
Estimating resolution as 143

real    0m1.933s
user    0m3.652s
sys     0m0.037s


$ time OMP_THREAD_LIMIT=2  tesseract -l eng  62771160-3bf82880-baa5-11e9-8d39-4d9c4381a093.png 1
Tesseract Open Source OCR Engine v4.0.0 with Leptonica
Warning: Invalid resolution 0 dpi. Using 70 instead.
Estimating resolution as 143

real    0m0.732s
user    0m0.757s
sys     0m0.032s

```

**Comment by ripefig** [comment]
@stweil Is there any solution? Maybe limit the default number of cores to 1 (or max cores - 1) until Tesseract can reliably work with multuple cores? Seems like it's completely broken for a lot of users and problem has persisted for years. This also breaks all the software that uses tesseract.

**Comment by stweil** [comment]
The right solution depends on your hardware (number of cores, memory interface) and your use case: on some hardware using more than one core results in faster OCR (see my results above), and training is much faster with 4 cores. It is always possible to either set `OMP_THREAD_LIMIT` or to build your own binary without multithreading. Without that, Tesseract is not "completely broken" or unreliable, but simply slow. I know that is not nice. The Windows binaries from UB Mannheim are therefore built without multithreading.

Because there are acceptable solutions for the speed issue, my current first priority is improving quality, not looking how to improve multithreading. If you or someone else finds a better solution for multithreading, a pull request would be welcomed.

**Comment by ripefig** [comment]
Out of the box, it takes about one hour to OCR a single page of text. It would take one month to OCR a textbook, and the CPU would probably fry. I think most users would consider this "completely broken," in the sense of not being usable.

The issue affects both AVX and non-AVX systems. The program is capable of cutting down times by two orders of magnitude in both cases, as demonstrated in this thread. Why not just limit the core count by default until the issue is fixed?

Of course, one could argue that it's up to application developers to make sure tesseract works on the target system. (I just tried a few OCR apps and most of them work fine - so it looks like they are fixing it on their end somehow).

**Comment by stweil** [comment]
That's simply not true. It is slow on your notebook. On my six year old notebook (no AVX) the official Debian package works pretty good:

    $ time tesseract 62841051-2b65cf00-bcac-11e9-97df-bf85ff0b09bf.png -
    Warning: Invalid resolution 0 dpi. Using 70 instead.
    Estimating resolution as 143
    
    real	0m0.327s
    user	0m0.796s
    sys	0m0.024s

**Comment by ripefig** [comment]
I didn't say it affects all systems, but it's frequent enough to warrant some kind of change. Multicore might result in 5-30% speed improvement but it can also result in a 10000% speed decrease on many systems. `Intel® Core™ i7-4600U CPU` isn't exactly an exotic chipset.

Perhaps you're saying this is an Ubuntu issue?

**Comment by amitdo** [comment]
>Processors: 4 × Intel® Core™ i7-4600U CPU @ 2.10GHz

https://ark.intel.com/content/www/us/en/ark/products/76616/intel-core-i7-4600u-processor-4m-cache-up-to-3-30-ghz.html

 >\# of Cores 2

**Comment by dagnelies** [comment]
Indeed, that whole multithreading thing caused more harm than good. There are a few issues around regarding this. I believe some people even compile specialized versions where OMP is completely removed since it runs more or less as fast, but with way less CPU consumption.

**Comment by Shreeshrii** [comment]
@stweil I want to compare the timing on Power8 to AVX2. I notice that the results you reported were with tesseract 4.0.0. Please rerun the test with the latest code.

My current result is:

```
 time tesseract -l eng 62841051-2b65cf00-bcac-11e9-97df-bf85ff0b09bf.png 1
Tesseract Open Source OCR Engine v5.0.0-alpha-537-g6f31 with Leptonica
Warning: Invalid resolution 0 dpi. Using 70 instead.
Estimating resolution as 143

real    0m1.109s
user    0m3.524s
sys     0m0.032s
```

**Comment by stweil** [comment]
Intermediate results (more results will get added later) with git master:

    # Power8, configure (default options)
    real	0m0.802s
    user	0m1.815s
    sys	0m0.030s

    # Power8, configure --disable-openmp --disable-shared
    real	0m1.243s
    user	0m1.231s
    sys	0m0.012s

**Comment by Shreeshrii** [comment]
@stweil Yes, I was testing with tessdata_best. It will be good if the results can be improved by using SIMD.

I tested with tessdata_fast just now. I get different timing results running the same command multiple times (there is lstmtraining running on the system). Is that to be expected?

I build with openmp as it is faster during training.

```
../configure  --enable-openmp --disable-debug --disable-opencl --disable-graphics   --disable-shared  --with-tensorflow=no 

tesseract 5.0.0-alpha-550-g944c1
 leptonica-1.78.0
  libgif 5.1.4 : libjpeg 8d (libjpeg-turbo 1.4.2) : libpng 1.2.54 : libtiff 4.0.9 : zlib 1.2.11 : libwebp 0.4.4 : libopenjp2 2.3.0

ubuntu@tesseract-ocr:~/TEST$ time tesseract -l eng 62841051-2b65cf00-bcac-11e9-97df-bf85ff0b09bf.png 1 --tessdata-dir ~/tessdata_fast
Tesseract Open Source OCR Engine v5.0.0-alpha-550-g944c1 with Leptonica
Warning: Invalid resolution 0 dpi. Using 70 instead.
Estimating resolution as 143

real    0m5.046s
user    0m9.020s
sys     0m0.112s


real    0m6.621s
user    0m11.644s
sys     0m0.126s


real    0m4.228s
user    0m7.744s
sys     0m0.111s


real    0m2.532s
user    0m4.124s
sys     0m0.053s


real    0m3.279s
user    0m5.673s
sys     0m0.086s


real    0m1.549s
user    0m2.538s
sys     0m0.053s


real    0m4.870s
user    0m8.351s
sys     0m0.118s


real    0m6.924s
user    0m13.265s
sys     0m0.165s
```

**Comment by Shreeshrii** [comment]
There is less variation in time taken with OMP_THREAD_LIMIT=1 

```
OMP_THREAD_LIMIT=1 time tesseract -l eng 62841051-2b65cf00-bcac-11e9-97df-bf85ff0b09bf.png 1 --tessdata-dir ~/tessdata_fast
Tesseract Open Source OCR Engine v5.0.0-alpha-550-g944c1 with Leptonica
Warning: Invalid resolution 0 dpi. Using 70 instead.
Estimating resolution as 143


1.25user 0.01system 0:01.28elapsed 99%CPU (0avgtext+0avgdata 36480maxresident)k
0inputs+128outputs (0major+800minor)pagefaults 0swaps

1.25user 0.01system 0:01.29elapsed 97%CPU (0avgtext+0avgdata 36416maxresident)k
0inputs+128outputs (0major+799minor)pagefaults 0swaps

1.24user 0.01system 0:01.27elapsed 99%CPU (0avgtext+0avgdata 36416maxresident)k
0inputs+128outputs (0major+799minor)pagefaults 0swaps

1.24user 0.02system 0:01.29elapsed 98%CPU (0avgtext+0avgdata 36416maxresident)k
0inputs+128outputs (0major+799minor)pagefaults 0swaps

1.24user 0.02system 0:01.31elapsed 96%CPU (0avgtext+0avgdata 36480maxresident)k
0inputs+128outputs (0major+800minor)pagefaults 0swaps

1.25user 0.01system 0:01.30elapsed 97%CPU (0avgtext+0avgdata 36416maxresident)k
0inputs+128outputs (0major+799minor)pagefaults 0swaps
```

**Comment by stweil** [comment]
As we both use the same environment (Power8 VM from OSU OSL, Ubuntu Bionic), our results should be comparable.

Of course a running training process will use much memory bandwidth and slow down a 2nd Tesseract process.

**Comment by stweil** [comment]
The training on Power8 should be much faster with SIMD. Here is the code which I used in my test:

    diff --git a/src/arch/dotproduct.cpp b/src/arch/dotproduct.cpp
    index 1a5e6991..91dcf9fe 100644
    --- a/src/arch/dotproduct.cpp
    +++ b/src/arch/dotproduct.cpp
    @@ -21,6 +21,11 @@ namespace tesseract {
     // Computes and returns the dot product of the two n-vectors u and v.
     double DotProductNative(const double* u, const double* v, int n) {
       double total = 0.0;
    +#ifdef _OPENMP
    +#pragma omp simd
    +#else
    +#error
    +#endif
       for (int k = 0; k < n; ++k) total += u[k] * v[k];
       return total;
     }

**Comment by Shreeshrii** [comment]
Thanks! I will give it a try. Do any compile options also need to be changed?

e.g. http://cache.freescale.com/files/training/doc/ftf/2014/FTF-NET-F0013.pdf shows

gcc -O3 -Wall -maltivec -mabi=altivec -mcpu=power8 -ftree-vectorize -ftree-vectorizer-verbose=5 -ffast-math -funsafe-math-optimizations ./timed_simple_dot_product.c -o ./timed_simple_dot_product.o

**Comment by Shreeshrii** [comment]
>As we both use the same environment (Power8 VM from OSU OSL, Ubuntu Bionic), our results should be comparable.

:-)

**Comment by stweil** [comment]
> Do any compile options also need to be changed?

Of course it needs OpenMP (otherwise the compiler will raise an error with the current patch), so `-fopenmp` is required for that file even if OpenMP was disabled for the rest of the code. I also added `-maltivec` and `-mabi=altivec` in my test. Maybe `-mcpu=native` could be added for all files.

**Comment by stweil** [comment]
New results for an ARMv8 based NVIDIA Xavier running Ubunto Bionic:

    # best, configure --disable-openmp --disable-shared
    real	0m5.502s
    user	0m5.352s
    sys	0m0.080s

    # best, configure --disable-openmp --disable-shared, SIMD
    real	0m3.534s
    user	0m3.400s
    sys	0m0.080s

**Comment by stweil** [comment]
The old ARM results were made with `tessdata_fast`. Here are new results:

    # best, configure --disable-openmp --disable-shared
    real	0m7.218s
    user	0m6.859s
    sys	0m0.248s

**Comment by Shreeshrii** [comment]
Can a different dot product calculation be used for Altivec - please see slide 16 in https://www.nxp.com/files-static/training_presentation/TP_ALTIVEC.pdf

**Comment by amitdo** [comment]
It's done automatically with the openmp-simd code.

**Comment by amitdo** [comment]
@stweil, did you benchmarked this against the manual code in a x86-64 machine?

**Comment by Shreeshrii** [comment]
>Of course it needs OpenMP (otherwise the compiler will raise an error with the current patch), so -fopenmp is required for that file even if OpenMP was disabled for the rest of the code. 

Shouldn't `--enable-openmp `  set `-fopenmp`?

I added the following to my build script.

```
export CXXFLAGS="-fopenmp -maltivec -mabi=altivec -mcpu=power8"
```

Now, `tesseract --version` reports about OPENMP - haven't seen it before with `--enable-openmp `  builds.

```
 tesseract -v
tesseract 5.0.0-alpha-554-g9ed3
 leptonica-1.78.0
  libgif 5.1.4 : libjpeg 8d (libjpeg-turbo 1.4.2) : libpng 1.2.54 : libtiff 4.0.9 : zlib 1.2.11 : libwebp 0.4.4 : libopenjp2 2.3.0
 Found OpenMP 201511
```

**Comment by amitdo** [comment]
>best, configure **--disable-openmp** --disable-shared

>Of course it needs OpenMP (otherwise the compiler will raise an error with the current patch), so -fopenmp is required for that file even if OpenMP was disabled for the rest of the code.

@stweil, It does not make sense to disable openmp and then to enable it.

**Comment by amitdo** [comment]
clang and gcc (>=4.9) both support the flag `-fopenmp-simd`.

**Comment by amitdo** [comment]
This code is more complete.

https://github.com/tesseract-ocr/tesseract/pull/983/commits/74f72e19b3185c558b39483f6e242e43838d60c5#diff-d0fa47c1b7e2cb742a89b8c8f824df62R343

**Comment by Shreeshrii** [comment]
|                        | fast<br>    default options                                      | fast<br>    OMP_THREAD_LIMIT=1                                   |
|------------------------|------------------------------------------------------------------|------------------------------------------------------------------|
| ripefig - intel        | real    1m9.096s<br>    user    3m7.484s<br>    sys     0m0.335s | real    0m0.366s<br>    user    0m0.346s<br>    sys     0m0.012s |
| stweil - intel         | real 0m0.327s<br>    user 0m0.796s<br>    sys 0m0.024s           |                                                                  |
| stweil<br>  -ARMv7     | real 0m1.864s<br>    user 0m4.778s<br>    sys 0m0.147s           | real 0m2.078s<br>    user 0m1.950s<br>    sys 0m0.099s           |
| stweil<br>  - ARMv8    |                                                                  |                                                                  |
| stweil - Power8        | real 0m0.802s<br>    user 0m1.815s<br>    sys 0m0.030s           |                                                                  |
| stweil - Recent Debian | real 0m0,209s<br>    user 0m0,497s<br>    sys 0m0,024s           | real 0m0,255s<br>    user 0m0,247s<br>    sys 0m0,008s           |

**Comment by Shreeshrii** [comment]
|       | best<br>    default options                            | best<br>    OMP_THREAD_LIMIT=1                         | best<br>    --disable-openmp<br>    --disable-shared           | best<br>    --disable-openmp<br>    --disable-shared<br>    SIMD |
|----------------------|--------------------------------------------------------|--------------------------------------------------------|----------------------------------------------------------------|------------------------------------------------------------------|
| ARMv7  |                                                        |                                                        | real 0m7.218s<br>    user 0m6.859s<br>    sys 0m0.248s<br>     | No Diff                                                          |
| ARMv8  |                                                        |                                                        | real 0m5.502s<br>    user 0m5.352s<br>    sys 0m0.080s<br>     | real 0m3.534s<br>    user 0m3.400s<br>    sys 0m0.080s<br>       |
| Power8      | real 0m1.329s<br>    user 0m3.804s<br>    sys 0m0.031s | real 0m3.155s<br>    user 0m3.139s<br>    sys 0m0.019s | real 0m2.981s<br>    user 0m2.957s<br>    sys 0m0.024s         | real 0m1.686s<br>    user 0m1.669s<br>    sys 0m0.016s           |

**Comment by Shreeshrii** [comment]
>The training on Power8 should be much faster with SIMD.

@stweil Which parts of training process will be speeded up by this? `lstmtraining`? I will like to test/benchmark with/without the suggested SIMD test patch.

**Comment by stweil** [comment]
Yes, `lstmtraining` will be faster, both for the training part and for the evaluation. That process uses up to two cores when OpenMP is disabled or up to 8 cores with OpenMP.

**Comment by Shreeshrii** [comment]
You had also mentioned earlier, in a different thread, about unrolled loops. Should that also be implemented along with this for Power?

https://github.com/tesseract-ocr/tesseract/pull/2106#issuecomment-464510169

**Comment by stweil** [comment]
Ideally loop unrolling should also be done by the compiler (try `-O3` or `-funroll-loops`).

**Comment by Shreeshrii** [comment]
I set both `-O3 -ffast-math` (similar to https://github.com/tesseract-ocr/tesseract/blob/2b688983f6c7abcce162559108d57ed159b616ea/CMakeLists.txt#L254 ) and the unittest `linlsq_test` failed with the following error. It works when I removed the `-ffast-math`.

```

Running main() from ../../unittest/../googletest/googletest/src/gtest_main.cc
[==========] Running 3 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 3 tests from LLSQTest
[ RUN      ] LLSQTest.BasicLines
[       OK ] LLSQTest.BasicLines (0 ms)
[ RUN      ] LLSQTest.Vectors
../../unittest/linlsq_test.cc:63: Failure
The difference between correct_vector.y() and vector.y() is 2, which exceeds tolerance, where
correct_vector.y() evaluates to 1,
vector.y() evaluates to -1, and
tolerance evaluates to 9.9999999747524271e-07.
[  FAILED  ] LLSQTest.Vectors (1 ms)
[ RUN      ] LLSQTest.RmsOrthWorksAsIntended
[       OK ] LLSQTest.RmsOrthWorksAsIntended (0 ms)
[----------] 3 tests from LLSQTest (1 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test suite ran. (1 ms total)
[  PASSED  ] 2 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] LLSQTest.Vectors
```

**Comment by Shreeshrii** [comment]
>lstmtraining will be faster, both for the training part and for the evaluation. That process uses up to two cores when OpenMP is disabled or up to 8 cores with OpenMP.

Is using up to 8 cores = 8 threads?

https://github.com/tesseract-ocr/tesseract/blob/944c1d9bffe59599924f3c04df4ae0c1e9852373/src/ccstruct/imagedata.h#L315

// A collection of DocumentData that knows roughly how much memory it is using.
// Note that while it supports background read-ahead, it assumes that a single
// thread is accessing documents, ie it is not safe for multiple threads to
// access different documents in parallel, as one may de-cache the other's
// content.

**Comment by stweil** [comment]
Yes, I should have written up to 8 threads. If there are only two CPUs with hyperthreading, those 8 threads will run on 4 cores, and the performance will be rather low. Of course you can use `OMP_THREAD_LIMIT=4` to handle this, but I am not sure how that will distribute the cores for training and evaluation.

The cache for image data works also with a separate thread, but that does not use OpenMP, so it also works when OpenMP was disabled.

**Comment by Shreeshrii** [comment]
>//ie it is not safe for multiple threads to
// access different documents in parallel, as one may de-cache the other's
// content.

I was asking about threads with regard to the above comment, whether multiple threads lead to slowing down.

**Comment by Shreeshrii** [comment]
| Tesstutorial<br>  Phases | Master                                                      | SIMD Patch                                                  | Performance      |
|--------------------------|-------------------------------------------------------------|-------------------------------------------------------------|------------------|
| 1-makedata               | real 87.50<br>    user 107.15<br>    sys 4.85<br>           | real 67.93<br>    user 107.51<br>    sys 4.88<br>           | SIMD real faster |
| 2-scratch                | real 7170.90<br>    user 21735.34<br>    sys 124.71<br>     | real 7244.33<br>    user 22014.08<br>    sys 128.46<br>     |                  |
| 3-impact-from-small      | real 643.20<br>    user 1893.29<br>    sys 12.16<br>        | real 654.52<br>    user 1936.23<br>    sys 14.80<br>        |                  |
| 4-impact-from-full       | real 1407.57<br>    user 4727.62<br>    sys 16.75<br>       | real 1464.04<br>    user 4887.35<br>    sys 20.09<br>       |                  |
| 5-makedata-plusminus     | real 91.26<br>    user 111.95<br>    sys 4.79<br>           | real 71.00<br>    user 124.34<br>    sys 5.82<br>           | SIMD real faster |
| 6-plusminus              | real 5975.63<br>    user 18346.83<br>    sys 60.97<br>      | real 7539.99<br>    user 20610.39<br>    sys 95.87<br>      | SIMD Worse       |
| 7-layer                  | real 2808.52<br>    user 8775.01<br>    sys 61.81<br>       | real 2793.53<br>    user 8661.41<br>    sys 51.00<br>       |                  |

**Comment by Shreeshrii** [comment]
I have posted above the results of my test on power8 running tesstutorial with tesseract built from git master vs with SIMD patch as suggested by @stweil.

The build was done using Advanced toolchain rather than the distro's gcc since 
> AT is highly recommended when you want to build an optimized CPU-bound application on POWER. ref: https://developer.ibm.com/linuxonpower/advance-toolchain/advtool-faq/

Build options in both cases included the following:

```
export CXXFLAGS="-O3 -maltivec -mabi=altivec -mcpu=power8 -mtune=power8 -fopenmp"

../../configure --enable-openmp --disable-debug --disable-opencl --disable-graphics --disable-shared --with-tensorflow=no 
```

**Comment by Shreeshrii** [comment]
Questions:

1. Is it OK to set the build options for all programs as I have done above?

2. I am planning also to test using @amitdo 's suggestion to use

` #pragma omp simd reduction(+:total)`

Should I expect the result to be very different from 

`` #pragma omp simd`

**Comment by Shreeshrii** [comment]
Updated training test results in https://github.com/tesseract-ocr/tesseract/issues/2611#issuecomment-558248658

**Comment by Danx69** [comment]
Tesseract was very slow running it within a script but I noticed it was very fast within a terminal then, using also your suggestions, I modify the script line in "OMP_THREAD_LIMIT=1 xterm -geometry 1X1+0+0 -e tesseract file1 file2" and obtained the same speed.
